File: 04-linked-lists/02_middle_list.py
import sys, os; sys.path.append(os.path.join(os.path.dirname(os.path.abspath(__file__)), '..'))
from test_runner import run_tests



# Middle of the Linked List (LC #876) -- Easy
# Input:  head: ListNode (non-empty)
# Output: ListNode (middle node; if two middles, return second)
# Return the middle node. If two middle nodes, return the second one.
#
# Example:
#   1->2->3->4->5       -> node 3  (middle of 5)
#   1->2->3->4->5->6    -> node 4  (second of two middles)
#   1                    -> node 1
#
#   Key: Slow/fast pointers. Slow moves 1 step, fast moves 2.
#        When fast reaches end, slow is at middle.

class ListNode:
    def __init__(self, val=0, nxt=None):
        self.val = val
        self.next = nxt


def make_list(arr):
    dummy = ListNode(0)
    curr = dummy
    for v in arr:
        curr.next = ListNode(v)
        curr = curr.next
    return dummy.next


def to_array(head):
    arr = []
    while head:
        arr.append(head.val)
        head = head.next
    return arr


def middle_node(head):
    pass


def solve(arr):
    node = middle_node(make_list(arr))
    return to_array(node)


run_tests(solve, [
    (([1,2,3,4,5],),     [3,4,5]),     # odd, return middle onward
    (([1,2,3,4,5,6],),   [4,5,6]),     # even, return second middle onward
    (([1],),              [1]),         # single node
    (([1,2],),            [2]),         # two nodes, return second
    (([1,2,3],),          [2,3]),       # three nodes
])

File: 04-linked-lists/02_middle_list.c
#include "../test_runner.h"

typedef struct ListNode {
    int val;
    struct ListNode* next;
} ListNode;

ListNode* make_node(int val) {
    ListNode* n = malloc(sizeof(ListNode));
    n->val = val; n->next = NULL;
    return n;
}

ListNode* make_list(int* arr, int n) {
    if (n == 0) return NULL;
    ListNode* head = make_node(arr[0]);
    ListNode* curr = head;
    for (int i = 1; i < n; i++) {
        curr->next = make_node(arr[i]);
        curr = curr->next;
    }
    return head;
}

void to_array(ListNode* head, int* out, int* len) {
    *len = 0;
    while (head) { out[(*len)++] = head->val; head = head->next; }
}


ListNode* middle_node(ListNode* head) {
    return NULL;
}

int main() {
    int out[10], len;
    to_array(middle_node(make_list(ARR(1,2,3,4,5), 5)), out, &len);
    TEST_ARR("odd",  out, ARR(3,4,5), 3);
    to_array(middle_node(make_list(ARR(1,2,3,4,5,6), 6)), out, &len);
    TEST_ARR("even", out, ARR(4,5,6), 3);
    TEST_SUMMARY();
}
